File: Java/com.faculdadevincit.designpatterns.observerpattern/src/Publisher.java
import java.util.ArrayList;
import java.util.List;

public class Publisher<TValue> {
    private List<IObservable<TValue>> observables;

    public Publisher() {
        observables = new ArrayList<>();
    }

    protected List<IObservable<TValue>> getObservables() {
        return observables;
    }

    public void addObservable(IObservable<TValue> observable) {
        observables.add(observable);
    }

    public void removeObservable(IObservable<TValue> observable) {
        observables.remove(observable);
    }

    public void notifyObservables(TValue value) {
        observables.forEach(o -> o.update(value));
    }
}

File: CSharp/FaculdadeVincit.DesignPatterns.ObserverPattern/Publisher.cs
namespace FaculdadeVincit.DesignPatterns.ObserverPattern;

public abstract class Publisher<TValue>
{
    protected List<IObservable<TValue>> Observables { get; private set; }

    public Publisher()
    {
        Observables = new List<IObservable<TValue>>();
    }

    public void AddObservable(IObservable<TValue> observable) => 
        Observables.Add(observable);

    public void RemoveObservable(IObservable<TValue> observable) =>
        Observables.Remove(observable);

    protected void NotifyObservables(TValue value) => 
        Observables.ForEach(o => o.Update(value));
}
